ice40/utils/fasm_icebox/fasm_icebox_utils.py

```diff
@@ -0,0 +1,541 @@
+"""Provide utilites for converting between FASM and icebox representations.
+
+FASM represents enabling features from a default "empty" configuration.
+IceBox documents what each bit does even if should be set by default (unused).
+
+Therefore any differences between ice40 parts are hidden from the FASM
+portion. For example IE, REN, and Ram PowerUp are inverted for 1k parts comparted to
+all other families.
+
+FASM features attempt to maintain the IceStrom nomenclature.
+Exceptions:
+ - '/' are replaced with '_' as '/' is not valid for FASM feature name
+ - RAM CBITS are combined into READ_MODE and WRITE_MODE FASM features
+ - IO configuration bits are combined into SimpleInput and SimpleOut FASM features
+"""
+
+from io import StringIO
+import re
+import unittest
+import numpy as np
+
+import icebox
+import fasm
+from ice40_feature import Feature, IceDbEntry, FasmEntry
+
+# hardcoded permutation
+# * https://github.com/YosysHQ/nextpnr/blob/343569105ddf7c97316922774dc4d70d1d4f7c9f/ice40/bitstream.cc#L460-L468
+# * https://github.com/cliffordwolf/icestorm/blob/master/icebox/icebox_hlc2asc.py#L925-L936
+# * https://github.com/YosysHQ/arachne-pnr/blob/840bdfdeb38809f9f6af4d89dd7b22959b176fdd/src/place.cc#L1573-L1627
+LUT_BITS = [4, 14, 15, 5, 6, 16, 17, 7, 3, 13, 12, 2, 1, 11, 10, 0]
+LUT_CTRL = {"CarryEnable": 8, "DffEnable": 9, "Set_NoReset": 18, "AsyncSetReset": 19}
+
+
+def _nibbles_to_bits(line):
+    """
+    Convert from icebox hex string for ramdata in asc files to an array of Bool
+    """
+    res = []
+    for ch in line:
+        res += [xx == "1" for xx in "{:4b}".format(int(ch, 16))]
+    res.reverse()
+    return res
+
+
+def _bits_to_nibbles(arr):
+    """
+    Convert from array of Bool to icebox hex string used for ramdata in asc files
+    """
+    res = []
+    for ii in range(0, len(arr), 4):
+        nibble_val = sum((1 << ii) for ii, xx in enumerate(arr[ii : ii + 4]) if xx)
+        res += "{:x}".format(nibble_val)
+    res.reverse()
+    return "".join(res)
+
+
+def _tile_to_array(tile, is_hex=False):
+    """
+    convert text icedb tile to a numpy array
+    """
+    if is_hex:
+        array = np.array([_nibbles_to_bits(line) for line in tile], dtype=bool)
+    else:
+        array = np.array([[int(xx) for xx in line] for line in tile], dtype=bool)
+    return array
+
+
+def _array_to_tile(tile_bits, tile, is_hex=False):
+    """
+    convert a numpy array to text icedb tile
```

Comment:
Docstring convention is to have the summary on the first line of the string:
https://www.python.org/dev/peps/pep-0257/

You have yours on the second line.